target/linux/lantiq/patches-5.10/0702-net-lantiq-enable-baby-jumbo-frames-on-xrx200.patch

```diff
@@ -0,0 +1,125 @@
+From d39d31486432bb3e54e4cb83744d99c3f5d773c6 Mon Sep 17 00:00:00 2001
+From: Thomas Nixon <tom@tomn.co.uk>
+Date: Sat, 3 Jul 2021 15:57:03 +0100
+Subject: [PATCH] net: lantiq: enable baby jumbo frames on xrx200
+
+xrx200 max MTU is reduced so that it works correctly when set to the
+max, and the max MTU of the switch is increased.
+
+The switch driver now enables non-standard MTUs on a per-port basis,
+with the overall frame size set based on the cpu port.
+
+When the MTU is not changed, this should have no effect.
+
+Signed-off-by: Thomas Nixon <tom@tomn.co.uk>
+---
+ drivers/net/dsa/lantiq_gswip.c       | 42 ++++++++++++++++++++++++++--
+ drivers/net/ethernet/lantiq_xrx200.c |  4 ++-
+ 2 files changed, 42 insertions(+), 4 deletions(-)
+
+diff --git a/drivers/net/dsa/lantiq_gswip.c b/drivers/net/dsa/lantiq_gswip.c
+index 93c7fa1fd4cb..1e67562153f9 100644
+--- a/drivers/net/dsa/lantiq_gswip.c
++++ b/drivers/net/dsa/lantiq_gswip.c
+@@ -238,6 +238,8 @@
+ 
+ #define XRX200_GPHY_FW_ALIGN	(16 * 1024)
+ 
++#define GSWIP_MAX_PACKET_LENGTH	9600
++
+ struct gswip_hw_info {
+ 	int max_ports;
+ 	int cpu_port;
+@@ -851,9 +853,6 @@ static int gswip_setup(struct dsa_switch *ds)
+ 	gswip_switch_mask(priv, 0, GSWIP_PCE_PCTRL_0_INGRESS,
+ 			  GSWIP_PCE_PCTRL_0p(cpu_port));
+ 
+-	gswip_switch_mask(priv, 0, GSWIP_MAC_CTRL_2_MLEN,
+-			  GSWIP_MAC_CTRL_2p(cpu_port));
+-	gswip_switch_w(priv, VLAN_ETH_FRAME_LEN + 8, GSWIP_MAC_FLEN);
+ 	gswip_switch_mask(priv, 0, GSWIP_BM_QUEUE_GCTRL_GL_MOD,
+ 			  GSWIP_BM_QUEUE_GCTRL);
+ 
+@@ -870,6 +869,8 @@ static int gswip_setup(struct dsa_switch *ds)
+ 		return err;
+ 	}
+ 
++	ds->mtu_enforcement_ingress = true;
++
+ 	gswip_port_enable(ds, cpu_port, NULL);
+ 	return 0;
+ }
+@@ -1426,6 +1427,39 @@ static int gswip_port_fdb_dump(struct dsa_switch *ds, int port,
+ 	return 0;
+ }
+ 
++static int gswip_port_max_mtu(struct dsa_switch *ds, int port)
++{
++	/* includes 8 bytes for special header */
++	return GSWIP_MAX_PACKET_LENGTH - VLAN_ETH_HLEN - ETH_FCS_LEN;
++}
++
++static int gswip_port_change_mtu(struct dsa_switch *ds, int port, int new_mtu)
++{
++	struct gswip_priv *priv = ds->priv;
++	int cpu_port = priv->hw_info->cpu_port;
++
++	/* cpu port always has maximum mtu of user ports, so use it to set
++	 * switch frame size, including 8 byte special header
++	 */
++	if (port == cpu_port) {
++		new_mtu += 8;
++		gswip_switch_w(priv, VLAN_ETH_HLEN + new_mtu + ETH_FCS_LEN,
++			       GSWIP_MAC_FLEN);
++	}
++
++	/* enable MLEN for ports with non-standard MTUs, including the special
++	 * header on the CPU port added above
++	 */
++	if (new_mtu != ETH_DATA_LEN)
++		gswip_switch_mask(priv, 0, GSWIP_MAC_CTRL_2_MLEN,
++				  GSWIP_MAC_CTRL_2p(port));
++	else
++		gswip_switch_mask(priv, GSWIP_MAC_CTRL_2_MLEN, 0,
++				  GSWIP_MAC_CTRL_2p(port));
```

Comment:
`GSWIP_MAC_CTRL_2_MLEN` has these options:
0: Normal (MTU of 1518 bytes ) VLAN tags are not included and can be added in addition.
1: Jumbo: the value from `GSWIP_MAC_FLEN` is used and it is with all VLAN tags. 

With `GSWIP_MAC_CTRL_2_LCHKL` you can also deactivate the length check. bit (2)

Reply:
Thanks, I hadn't seen that. Do you think it would make more sense to disable the length check on the CPU port so that the special header does not affect the maximum packet size on user ports with non-standard MTUs, as it does currently?

Reply:
from what I understand DSA includes the tag header size ("tag protocol overhead") in the MTU calculation for the CPU port:
https://elixir.bootlin.com/linux/v5.15-rc2/source/net/dsa/slave.c#L1640

Reply:
I would prefer if we still check the MTU on the CPU port we should at least make sure the packets are fitting into the DMA descriptors, I do not know what the hardware will do if the DMA descriptors are too small.

Reply:
@xdarklight The tag overhead is not included in the CPU port MTU, but it is included in the master port MTU, so i think this logic is correct.

Adding a printk to gswip_port_change_mtu:

```c
printk("gswip_port_change_mtu %d %d %d\n", port, cpu_port, new_mtu);
```

results in this on boot up:
```
[   10.260315] gswip_port_change_mtu 6 6 1500
[   10.262955] gswip_port_change_mtu 0 6 1500
[   10.283266] gswip_port_change_mtu 1 6 1500
[   10.303300] gswip_port_change_mtu 2 6 1500
[   10.322546] gswip_port_change_mtu 4 6 1500
[   10.342211] gswip_port_change_mtu 5 6 1500
```

And this when running `ip link set dev lan1 mtu 1508`:
```
[  144.866115] gswip_port_change_mtu 6 6 1508
[  144.868820] gswip_port_change_mtu 4 6 1508
[  144.873023] gswip_port_change_mtu 2 6 1508
[  144.876964] gswip_port_change_mtu 1 6 1508
[  144.881200] gswip_port_change_mtu 0 6 1508
```

Reply:
@hauke

> I would prefer if we still check the MTU on the CPU port we should at least make sure the packets are fitting into the DMA descriptors, I do not know what the hardware will do if the DMA descriptors are too small.

The DSA infrastructure doesn't allow the slave device mtu to be set higher than the maximum mtu of the master device (minus the overhead), so with the current patch I don't think this is a problem.

Unfortunately it is possible to reduce the master MTU below that required by the slaves, e.g.:

```bash
ip link set dev lan1 mtu 1508 # master: 1516, slaves: 1508
ip link set dev eth0 mtu 1500 # master: 1500, slaves: 1508
```

In theory this might cause an issue, but the [xrx200 jumbo frame patch](https://git.kernel.org/pub/scm/linux/kernel/git/netdev/net-next.git/commit/?id=998ac358019e491217e752bc6dcbb3afb2a6fa3e) from @abajk never reduces the buffer sizes, so it should be OK. I've tried this and couldn't see any issues, but maybe there's a better way that I'm missing?

Reply:
> In theory this might cause an issue, but the [xrx200 jumbo frame patch](https://git.kernel.org/pub/scm/linux/kernel/git/netdev/net-next.git/commit/?id=998ac358019e491217e752bc6dcbb3afb2a6fa3e) from @abajk never reduces the buffer sizes, so it should be OK. I've tried this and couldn't see any issues, but maybe there's a better way that I'm missing?

My patch does not reallocate buffers if the new MTU is less than the old one. Having a larger buffer, the driver can receive a shorter packet. If the driver receives a packet, the next buffer buffer will be smaller. After 64 packets are received, all buffers will be smaller.

Reply:
> OK. I've tried this and couldn't see any issues, but maybe there's a better way that I'm missing?

It means that large packets on your device do not fhangs the switch?

Reply:
> > In theory this might cause an issue, but the [xrx200 jumbo frame patch](https://git.kernel.org/pub/scm/linux/kernel/git/netdev/net-next.git/commit/?id=998ac358019e491217e752bc6dcbb3afb2a6fa3e) from @abajk never reduces the buffer sizes, so it should be OK. I've tried this and couldn't see any issues, but maybe there's a better way that I'm missing?
> 
> My patch does not reallocate buffers if the new MTU is less than the old one. Having a larger buffer, the driver can receive a shorter packet. If the driver receives a packet, the next buffer buffer will be smaller. After 64 packets are received, all buffers will be smaller.

Ah right, thanks for clarifying that.

> > OK. I've tried this and couldn't see any issues, but maybe there's a better way that I'm missing?
> 
> It means that large packets on your device do not fhangs the switch?

Sorry, i meant issues with reducing the master MTU below the mtu required by the slaves.

I was not testing with enough packets, but trying it again it just seems to drop packets, which should be fine.

I see the same issue as you with large packets hanging the switch -- assuming it is a problem with the switch, GSWIP_MAX_PACKET_LENGTH should probably be reduced.